File: integration-sample/php/central_logs.php
<?php
/**
 * central-logs integration client — batched NDJSON inserts (PHP 7.4+).
 *
 * Setup (see INTEGRATION.md):
 *   1. Admin key pinned in the server's .env (CENTRAL_LOGS_HTTP_API_KEY).
 *   2. Mint an insert-only key (POST /v1/api-keys, scopes: "insert").
 *   3. Export for this app:
 *        CENTRAL_LOGS_URL      (default http://localhost:8080)
 *        CENTRAL_LOGS_API_KEY  (the raw clk_... value — required)
 *        CENTRAL_LOGS_SERVICE  (default "default-app")
 *
 * Usage:
 *   require_once __DIR__ . '/central_logs.php';
 *   central_logs_log('info', 'app started', ['version' => '1.2.3']);
 *   central_logs_log('warn', 'queue lag 12s', ['queue' => 'events', 'duration_ms' => 12000]);
 *   // registered_shutdown_function flushes automatically — nothing to call.
 *
 * PHP has no background timer in classic SAPIs, so batches flush when
 * MAX_BATCH is reached or at request shutdown. For long-running CLI daemons
 * call central_logs_flush() on your own loop cadence.
 */

declare(strict_types=1);

final class CentralLogsClient
{
    private string $url;
    private string $apiKey;
    private string $service;
    private int $maxBatch;
    private int $maxQueue;
    /** @var list<array<string,mixed>> */
    private array $queue = [];
    private int $dropped = 0;

    public function __construct(
        ?string $url = null,
        ?string $apiKey = null,
        ?string $service = null
    ) {
        $this->url      = rtrim($url      ?? (getenv('CENTRAL_LOGS_URL')      ?: 'http://localhost:8080'), '/');
        $this->apiKey   = $apiKey         ?? (getenv('CENTRAL_LOGS_API_KEY')  ?: '');
        $this->service  = $service        ?? (getenv('CENTRAL_LOGS_SERVICE')  ?: 'default-app');
        $this->maxBatch = (int) (getenv('CENTRAL_LOGS_MAX_BATCH') ?: '100');
        $this->maxQueue = (int) (getenv('CENTRAL_LOGS_MAX_QUEUE') ?: '5000');

        if (function_exists('register_shutdown_function')) {
            register_shutdown_function(fn () => $this->flush());
        }
    }

    /** Queue one record. Extra key/value pairs become queryable attributes. */
    public function log(string $level, string $msg, array $fields = []): void
    {
        $record = array_merge([
            'service' => $this->service,
            'level'   => strtolower($level),
            'msg'     => $msg,
            'ts'      => gmdate('Y-m-d\TH:i:s.v\Z'),
        ], $fields);

        if (count($this->queue) >= $this->maxQueue) {
            array_shift($this->queue); // bounded: drop oldest
        }
        $this->queue[] = $record;

        if (count($this->queue) >= $this->maxBatch) {
            $this->flush();
        }
    }

    /** Send the queued batch now. */
    public function flush(): void
    {
        if (!$this->queue) {
            return;
        }
        $batch = $this->queue;
        $this->queue = [];

        if ($this->apiKey === '') {
            $this->drop($batch, 'CENTRAL_LOGS_API_KEY not set');
            return;
        }

        $body = '';
        foreach ($batch as $record) {
            $body .= json_encode($record, JSON_UNESCAPED_SLASHES) . "\n";
        }

        $ch = curl_init($this->url . '/v1/logs');
        curl_setopt_array($ch, [
            CURLOPT_POST           => true,
            CURLOPT_POSTFIELDS     => $body,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT        => 5,
            CURLOPT_HTTPHEADER     => [
                'Content-Type: application/x-ndjson',
                'Authorization: Bearer ' . $this->apiKey,
            ],
        ]);
        $resp = curl_exec($ch);
        if ($resp === false) {
            $this->drop($batch, 'curl: ' . curl_error($ch));
        } else {
            $status = (int) curl_getinfo($ch, CURLINFO_RESPONSE_CODE);
            if ($status !== 200) {
                $this->drop($batch, "HTTP $status: " . substr((string) $resp, 0, 200));
            } else {
                $out = json_decode((string) $resp, true);
                if (is_array($out) && ($out['rejected'] ?? 0) > 0) {
                    error_log(sprintf(
                        '[central-logs] rejected %d/%d records',
                        (int) $out['rejected'],
                        count($batch)
                    ));
                }
            }
        }
        curl_close($ch);
    }

    /** @param list<array<string,mixed>> $batch */
    private function drop(array $batch, string $reason): void
    {
        $this->dropped += count($batch);
        error_log("[central-logs] flush failed ({$this->dropped} dropped total): $reason");
        foreach ($batch as $record) {
            error_log(json_encode($record, JSON_UNESCAPED_SLASHES));
        }
    }
}

/** Shared default instance. */
$GLOBALS['__central_logs'] = $GLOBALS['__central_logs'] ?? new CentralLogsClient();

function central_logs_log(string $level, string $msg, array $fields = []): void
{
    $GLOBALS['__central_logs']->log($level, $msg, $fields);
}

function central_logs_flush(): void
{
    $GLOBALS['__central_logs']->flush();
}

// ── smoke test ─────────────────────────────────────────────────────────────
//   CENTRAL_LOGS_API_KEY=clk_... php central_logs.php
if (realpath($argv[0] ?? '') === __FILE__) {
    central_logs_log('info', 'central-logs php client smoke test', ['env' => 'test']);
    central_logs_flush();
    echo "flushed; verify with: curl '.../api/logs?filter=service:default-app&window=5m'\n";
}

File: integration-sample/python/central_logs.py
#!/usr/bin/env python3
"""central-logs integration client — batched NDJSON inserts.

Setup (see INTEGRATION.md):
    1. Admin key pinned in the server's .env (CENTRAL_LOGS_HTTP_API_KEY).
    2. Mint an insert-only key:
         curl -sf -X POST http://localhost:8080/v1/api-keys \\
           -H "Authorization: Bearer $ADMIN_KEY" \\
           -H 'Content-Type: application/json' \\
           -d '{"name":"my-app","scopes":"insert"}'
    3. Export for this app:
         CENTRAL_LOGS_URL      (default http://localhost:8080)
         CENTRAL_LOGS_API_KEY  (the raw clk_... value — required)
         CENTRAL_LOGS_SERVICE  (default "default-app")

Usage:
    from central_logs import log          # shared instance
    log("info", "app started", version="1.2.3")
    log("warn", "queue lag 12s", queue="events", duration_ms=12000)

    # or an explicitly configured instance
    from central_logs import CentralLogs
    cl = CentralLogs(service="payments")
    cl.log("error", "gateway timeout", duration_ms=3000)
    cl.close()                            # final flush

Env vars are re-read lazily so dotenv-style imports work. On persistent
ingest failure batches go to stderr (nothing silently vanishes); the queue
is bounded so a long outage cannot exhaust memory.
"""

from __future__ import annotations

import atexit
import json
import os
import sys
import threading
from datetime import datetime, timezone
from typing import Any
from urllib import error as urlerror
from urllib import request as urlrequest

_LEVELS = ("debug", "info", "warn", "error", "fatal")


class CentralLogs:
    def __init__(
        self,
        url: str | None = None,
        api_key: str | None = None,
        service: str | None = None,
        flush_ms: int | None = None,
        max_batch: int | None = None,
        max_queue: int | None = None,
    ) -> None:
        self._url = (url or os.environ.get("CENTRAL_LOGS_URL") or "http://localhost:8080").rstrip("/")
        self._api_key = api_key if api_key is not None else os.environ.get("CENTRAL_LOGS_API_KEY", "")
        self._service = service or os.environ.get("CENTRAL_LOGS_SERVICE", "default-app")
        self._flush_ms = flush_ms or int(os.environ.get("CENTRAL_LOGS_FLUSH_MS", "1000"))
        self._max_batch = max_batch or int(os.environ.get("CENTRAL_LOGS_MAX_BATCH", "100"))
        self._max_queue = max_queue or int(os.environ.get("CENTRAL_LOGS_MAX_QUEUE", "5000"))

        self._queue: list[dict[str, Any]] = []
        self._lock = threading.Lock()
        self._dropped = 0
        self._closed = False

        self._wake = threading.Event()
        self._worker = threading.Thread(target=self._run, name="central-logs", daemon=True)
        self._worker.start()
        atexit.register(self.close)

    # ── public API ─────────────────────────────────────────────────────────
    def log(self, level: str, msg: str, **fields: Any) -> None:
        """Queue one record. `fields` become queryable attributes."""
        level = level.lower()
        if level not in _LEVELS:
            level = "info"
        record = {
            "service": self._service,
            "level": level,
            "msg": msg,
            "ts": datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z"),
            **fields,
        }
        with self._lock:
            if len(self._queue) >= self._max_queue:
                self._queue.pop(0)  # bounded: drop oldest
            self._queue.append(record)
            full = len(self._queue) >= self._max_batch
        if full:
            self.flush()

    def flush(self, timeout: float = 5.0) -> None:
        """Send the queued batch now. Safe to call concurrently."""
        batch: list[dict[str, Any]] = []
        with self._lock:
            if self._queue:
                batch, self._queue = self._queue[:], []
        if not batch:
            return
        if not self._api_key:
            self._drop(batch, "CENTRAL_LOGS_API_KEY not set")
            return
        body = "\n".join(json.dumps(r, default=str) for r in batch)
        try:
            req = urlrequest.Request(
                f"{self._url}/v1/logs",
                data=body.encode("utf-8"),
                method="POST",
                headers={
                    "Content-Type": "application/x-ndjson",
                    "Authorization": f"Bearer {self._api_key}",
                },
            )
            with urlrequest.urlopen(req, timeout=timeout) as resp:
                out = json.loads(resp.read().decode("utf-8") or "{}")
                rejected = out.get("rejected", 0)
                if rejected:
                    print(f"[central-logs] rejected {rejected}/{len(batch)} records", file=sys.stderr)
        except urlerror.HTTPError as e:
            self._drop(batch, f"HTTP {e.code}: {e.read()[:200]!r}")
        except Exception as e:  # noqa: BLE001 — network layer, best-effort
            self._drop(batch, str(e))

    def close(self) -> None:
        """Stop the worker and flush whatever is left."""
        if self._closed:
            return
        self._closed = True
        self._wake.set()
        self._worker.join(timeout=3)
        self.flush()

    # ── internals ──────────────────────────────────────────────────────────
    def _drop(self, batch: list[dict[str, Any]], reason: str) -> None:
        self._dropped += len(batch)
        print(f"[central-logs] flush failed ({self._dropped} dropped total): {reason}", file=sys.stderr)
        for r in batch:
            print(json.dumps(r, default=str), file=sys.stderr)

    def _run(self) -> None:
        while not self._closed:
            self._wake.wait(self._flush_ms / 1000.0)
            self._wake.clear()
            if self._closed:
                break
            self.flush()


# Module-level shared instance + convenience function.
_default = CentralLogs()


def log(level: str, msg: str, **fields: Any) -> None:
    _default.log(level, msg, **fields)


if __name__ == "__main__":
    # Smoke test:  CENTRAL_LOGS_API_KEY=... python3 central_logs.py
    log("info", "central-logs python client smoke test", env="test")
    _default.flush()
    print("flushed; verify with: curl '.../api/logs?filter=service:<service>&window=5m'")
